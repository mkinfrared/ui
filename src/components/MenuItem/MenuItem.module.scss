@use "../../styles/common.scss" as v;

.MenuItem {
  background-color: var(--background);
  cursor: pointer;
  font-size: 1.6rem;
  list-style: none;
  padding: 1rem 1.6rem;
  transition: background-color 250ms var(--ease-4);

  &[aria-selected="true"] {
    --background: var(--purple-gradient);
  }

  &:hover {
    --background: var(--purple-light);
  }

  &:focus-visible {
    @include v.focus-visible;
  }
}
