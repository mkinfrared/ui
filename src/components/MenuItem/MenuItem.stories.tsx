import { Meta, Story } from "@storybook/react";

import MenuItem from "./MenuItem";
import { MenuItemProps } from "./MenuItem.type";
import css from "./Story.module.scss";

export default {
  title: "UI/MenuItem",
  component: MenuItem,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const Template: Story<MenuItemProps<any>> = (args) => (
  <div className={css.Story}>
    <MenuItem {...args} />
  </div>
);

const Default = Template.bind({});

Default.args = {};

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default };
