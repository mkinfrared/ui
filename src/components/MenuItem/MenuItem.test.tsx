import { fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import MenuItem from "./MenuItem";

describe("<MenuItem />", () => {
  const onClick = jest.fn();
  const onFocus = jest.fn();

  const Component = (
    <>
      <MenuItem onClick={onClick} onFocus={onFocus} value={42} isSelected>
        Value 42
      </MenuItem>
      <MenuItem onClick={onClick} onFocus={onFocus} value={43}>
        Value 43
      </MenuItem>
    </>
  );

  beforeAll(() => {
    window.HTMLElement.prototype.scrollIntoView = () => null;
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should be defined", () => {
    expect(MenuItem).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getAllByTestId } = render(Component);
    const elements = getAllByTestId("MenuItem");

    expect(elements.length).toBe(2);
  });

  it("should call 'onClick' from props on item click event", () => {
    const { getAllByTestId } = render(Component);
    const [element] = getAllByTestId("MenuItem");

    fireEvent.click(element);

    expect(onClick).toHaveBeenCalledTimes(1);

    expect(onClick).toHaveBeenCalledWith(42);
  });

  it("should call 'onClick' from props on item 'Enter' key down", async () => {
    const { getAllByTestId } = render(Component);
    const [element] = getAllByTestId("MenuItem");

    await userEvent.tab();

    fireEvent.keyDown(element, { key: "Enter" });

    expect(onClick).toHaveBeenCalledTimes(1);

    expect(onClick).toHaveBeenCalledWith(42);
  });

  it("should call 'onFocus' from props on item focus event", () => {
    const { getAllByTestId } = render(Component);
    const elements = getAllByTestId("MenuItem");
    const [element] = elements;

    fireEvent.focus(element);

    expect(onFocus).toHaveBeenCalledTimes(1);

    expect(onFocus).toHaveBeenCalledWith(42);
  });

  it("should move focus to next menu item on 'ArrowDown' key press", async () => {
    const { getAllByTestId } = render(Component);
    const elements = getAllByTestId("MenuItem");
    const [first, second] = elements;

    await userEvent.tab();

    expect(first).toHaveAttribute("tabindex", "0");

    expect(second).toHaveAttribute("tabindex", "-1");

    fireEvent.keyDown(first, { key: "ArrowDown" });

    expect(first).toHaveAttribute("tabindex", "-1");

    expect(second).toHaveAttribute("tabindex", "0");
  });

  it("should not move focus from the item on 'ArrowDown' key press if it's the last in list", async () => {
    const { getAllByTestId } = render(Component);
    const elements = getAllByTestId("MenuItem");
    const [first, second] = elements;

    await userEvent.tab();

    expect(first).toHaveAttribute("tabindex", "0");

    expect(second).toHaveAttribute("tabindex", "-1");

    fireEvent.keyDown(first, { key: "ArrowDown" });

    expect(first).toHaveAttribute("tabindex", "-1");

    expect(second).toHaveAttribute("tabindex", "0");

    fireEvent.keyDown(second, { key: "ArrowDown" });

    expect(first).toHaveAttribute("tabindex", "-1");

    expect(second).toHaveAttribute("tabindex", "0");
  });

  it("should move focus to previous menu item on 'ArrowUp' key press", async () => {
    const { getAllByTestId } = render(Component);
    const elements = getAllByTestId("MenuItem");
    const [first, second] = elements;

    await userEvent.tab();

    expect(first).toHaveAttribute("tabindex", "0");

    expect(second).toHaveAttribute("tabindex", "-1");

    fireEvent.keyDown(first, { key: "ArrowDown" });

    expect(first).toHaveAttribute("tabindex", "-1");

    expect(second).toHaveAttribute("tabindex", "0");

    fireEvent.keyDown(second, { key: "ArrowUp" });

    expect(first).toHaveAttribute("tabindex", "0");

    expect(second).toHaveAttribute("tabindex", "-1");
  });

  it("should not move focus from the item on 'ArrowUp' key press if it's the first in list", async () => {
    const { getAllByTestId } = render(Component);
    const elements = getAllByTestId("MenuItem");
    const [first, second] = elements;

    await userEvent.tab();

    expect(first).toHaveAttribute("tabindex", "0");

    expect(second).toHaveAttribute("tabindex", "-1");

    fireEvent.keyDown(first, { key: "ArrowUp" });

    expect(first).toHaveAttribute("tabindex", "0");

    expect(second).toHaveAttribute("tabindex", "-1");
  });
});
