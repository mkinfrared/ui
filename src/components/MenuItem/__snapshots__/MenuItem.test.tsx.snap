// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuItem /> should match the snapshot 1`] = `
<div>
  <li
    aria-selected="true"
    class="MenuItem"
    data-testid="MenuItem"
    role="option"
    tabindex="0"
  >
    Value 42
  </li>
  <li
    class="MenuItem"
    data-testid="MenuItem"
    role="option"
    tabindex="-1"
  >
    Value 43
  </li>
</div>
`;
